// mattermost-redux/lib/selectors/entities/properties.js
"use strict";
// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPropertyValuesForField = exports.getPropertyValuesForTargetByFieldIds = exports.getPropertyValuesForTarget = exports.getPropertyFieldsByIds = exports.getPropertyFieldsForObjectTypeAndGroup = void 0;
exports.getPropertyFieldById = getPropertyFieldById;
exports.getPropertyGroupById = getPropertyGroupById;
exports.getPropertyGroupByName = getPropertyGroupByName;
exports.getPropertyValueForTargetField = getPropertyValueForTargetField;
const create_selector_1 = require("mattermost-redux/selectors/create_selector");
// Field selectors
function getPropertyFieldsById(state) {
    return state.entities.properties.fields.byId;
}
exports.getPropertyFieldsForObjectTypeAndGroup = (0, create_selector_1.createSelector)('getPropertyFieldsForObjectTypeAndGroup', (state, objectType, groupId) => state.entities.properties.fields.byObjectType[objectType]?.[groupId], (fields) => {
    if (!fields) {
        return [];
    }
    return Object.values(fields);
});
function getPropertyFieldById(state, fieldId) {
    return getPropertyFieldsById(state)[fieldId];
}
exports.getPropertyFieldsByIds = (0, create_selector_1.createSelector)('getPropertyFieldsByIds', getPropertyFieldsById, (state, fieldIds) => fieldIds, (byId, fieldIds) => {
    return fieldIds.reduce((acc, id) => {
        const field = byId[id];
        if (field) {
            acc.push(field);
        }
        return acc;
    }, []);
});
// Group selectors
function getPropertyGroupById(state, groupId) {
    return state.entities.properties.groups.byId[groupId];
}
function getPropertyGroupByName(state, name) {
    return state.entities.properties.groups.byName[name];
}
// Value selectors
exports.getPropertyValuesForTarget = (0, create_selector_1.createSelector)('getPropertyValuesForTarget', (state, targetId) => state.entities.properties.values.byTargetId[targetId], (targetValues) => {
    if (!targetValues) {
        return [];
    }
    return Object.values(targetValues);
});
function getPropertyValueForTargetField(state, targetId, fieldId) {
    return state.entities.properties.values.byTargetId[targetId]?.[fieldId];
}
exports.getPropertyValuesForTargetByFieldIds = (0, create_selector_1.createSelector)('getPropertyValuesForTargetByFieldIds', (state, targetId) => state.entities.properties.values.byTargetId[targetId], (state, targetId, fieldIds) => fieldIds, (targetValues, fieldIds) => {
    if (!targetValues) {
        return [];
    }
    return fieldIds.reduce((acc, fieldId) => {
        const value = targetValues[fieldId];
        if (value) {
            acc.push(value);
        }
        return acc;
    }, []);
});
exports.getPropertyValuesForField = (0, create_selector_1.createSelector)('getPropertyValuesForField', (state, fieldId) => state.entities.properties.values.byFieldId[fieldId], (fieldValues) => {
    if (!fieldValues) {
        return [];
    }
    return Object.values(fieldValues);
});
