// mattermost-redux/lib/actions/access_control.js
"use strict";
// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAccessControlPolicy = getAccessControlPolicy;
exports.createAccessControlPolicy = createAccessControlPolicy;
exports.deleteAccessControlPolicy = deleteAccessControlPolicy;
exports.searchAccessControlPolicies = searchAccessControlPolicies;
exports.searchPermissionPolicies = searchPermissionPolicies;
exports.searchTeamAccessControlPolicies = searchTeamAccessControlPolicies;
exports.searchAccessControlPolicyChannels = searchAccessControlPolicyChannels;
exports.assignChannelsToAccessControlPolicy = assignChannelsToAccessControlPolicy;
exports.unassignChannelsFromAccessControlPolicy = unassignChannelsFromAccessControlPolicy;
exports.getAccessControlFields = getAccessControlFields;
exports.searchUsersForExpression = searchUsersForExpression;
exports.getVisualAST = getVisualAST;
exports.validateExpressionAgainstRequester = validateExpressionAgainstRequester;
exports.createAccessControlSyncJob = createAccessControlSyncJob;
exports.updateAccessControlPoliciesActive = updateAccessControlPoliciesActive;
const redux_batched_actions_1 = require("redux-batched-actions");
const action_types_1 = require("mattermost-redux/action_types");
const client_1 = require("mattermost-redux/client");
const helpers_1 = require("./helpers");
function getAccessControlPolicy(id, channelId, teamId) {
    return (0, helpers_1.bindClientFunc)({
        clientFunc: () => client_1.Client4.getAccessControlPolicy(id, channelId, teamId),
        onSuccess: [action_types_1.AdminTypes.RECEIVED_ACCESS_CONTROL_POLICY],
        params: [],
    });
}
function createAccessControlPolicy(policy, teamId) {
    return async (dispatch, getState) => {
        let data;
        try {
            data = await client_1.Client4.updateOrCreateAccessControlPolicy(policy, teamId);
        }
        catch (error) {
            (0, helpers_1.forceLogoutIfNecessary)(error, dispatch, getState);
            return { error };
        }
        dispatch({ type: action_types_1.AdminTypes.CREATE_ACCESS_CONTROL_POLICY_SUCCESS, data });
        return { data };
    };
}
function deleteAccessControlPolicy(id, teamId) {
    return (0, helpers_1.bindClientFunc)({
        clientFunc: () => client_1.Client4.deleteAccessControlPolicy(id, teamId),
        onSuccess: [action_types_1.AdminTypes.DELETE_ACCESS_CONTROL_POLICY_SUCCESS],
        params: [],
    });
}
function searchAccessControlPolicies(term, type, after, limit) {
    return async (dispatch, getState) => {
        let data;
        try {
            data = await client_1.Client4.searchAccessControlPolicies(term, type, after, limit);
        }
        catch (error) {
            (0, helpers_1.forceLogoutIfNecessary)(error, dispatch, getState);
            return { error };
        }
        dispatch({ type: action_types_1.AdminTypes.RECEIVED_ACCESS_CONTROL_POLICIES_SEARCH, data: data.policies });
        return { data };
    };
}
function searchPermissionPolicies(term, after, limit) {
    return async (dispatch, getState) => {
        let data;
        try {
            data = await client_1.Client4.searchPermissionPolicies(term, after, limit);
        }
        catch (error) {
            (0, helpers_1.forceLogoutIfNecessary)(error, dispatch, getState);
            return { error };
        }
        dispatch({ type: action_types_1.AdminTypes.RECEIVED_ACCESS_CONTROL_POLICIES_SEARCH, data: data.policies });
        return { data };
    };
}
function searchTeamAccessControlPolicies(teamId, term, type, after, limit) {
    return async (dispatch, getState) => {
        if (!teamId) {
            return { error: new Error('teamId is required for team-scoped policy search') };
        }
        let data;
        try {
            data = await client_1.Client4.searchAccessControlPolicies(term, type, after, limit, teamId);
        }
        catch (error) {
            (0, helpers_1.forceLogoutIfNecessary)(error, dispatch, getState);
            return { error };
        }
        dispatch({ type: action_types_1.AdminTypes.RECEIVED_ACCESS_CONTROL_POLICIES_SEARCH, data: data.policies });
        return { data };
    };
}
function searchAccessControlPolicyChannels(id, term, opts, teamId) {
    return async (dispatch, getState) => {
        let data;
        try {
            data = await client_1.Client4.searchChildAccessControlPolicyChannels(id, term, opts, teamId);
        }
        catch (error) {
            (0, helpers_1.forceLogoutIfNecessary)(error, dispatch, getState);
            return { error };
        }
        const childs = {};
        childs[id] = data.channels.map((channel) => channel.id);
        dispatch((0, redux_batched_actions_1.batchActions)([
            { type: action_types_1.AdminTypes.RECEIVED_ACCESS_CONTROL_CHILD_POLICIES, data: childs },
            { type: action_types_1.ChannelTypes.RECEIVED_CHANNELS, data: data.channels },
        ]));
        return { data };
    };
}
function assignChannelsToAccessControlPolicy(policyId, channelIds, teamId) {
    return (0, helpers_1.bindClientFunc)({
        clientFunc: () => client_1.Client4.assignChannelsToAccessControlPolicy(policyId, channelIds, teamId),
        onSuccess: [action_types_1.AdminTypes.ASSIGN_CHANNELS_TO_ACCESS_CONTROL_POLICY_SUCCESS],
        params: [],
    });
}
function unassignChannelsFromAccessControlPolicy(policyId, channelIds, teamId) {
    return (0, helpers_1.bindClientFunc)({
        clientFunc: () => client_1.Client4.unassignChannelsFromAccessControlPolicy(policyId, channelIds, teamId),
        onSuccess: [action_types_1.AdminTypes.UNASSIGN_CHANNELS_FROM_ACCESS_CONTROL_POLICY_SUCCESS],
        params: [],
    });
}
function getAccessControlFields(after, limit, channelId, teamId) {
    return (0, helpers_1.bindClientFunc)({
        clientFunc: () => client_1.Client4.getAccessControlFields(after, limit, channelId, teamId),
        params: [],
    });
}
function searchUsersForExpression(expression, term, after, limit, channelId, teamId) {
    return async (dispatch, getState) => {
        let data;
        try {
            data = await client_1.Client4.testAccessControlExpression(expression, term, after, limit, channelId, teamId);
        }
        catch (error) {
            (0, helpers_1.forceLogoutIfNecessary)(error, dispatch, getState);
            return { error };
        }
        dispatch({ type: action_types_1.UserTypes.RECEIVED_PROFILES, data: data.users });
        return { data };
    };
}
function getVisualAST(expression, channelId, teamId) {
    return (0, helpers_1.bindClientFunc)({
        clientFunc: () => client_1.Client4.expressionToVisualFormat(expression, channelId, teamId),
        params: [],
    });
}
function validateExpressionAgainstRequester(expression, channelId, teamId) {
    return async (dispatch, getState) => {
        let data;
        try {
            data = await client_1.Client4.validateExpressionAgainstRequester(expression, channelId, teamId);
        }
        catch (error) {
            (0, helpers_1.forceLogoutIfNecessary)(error, dispatch, getState);
            return { error };
        }
        return { data };
    };
}
function createAccessControlSyncJob(jobData) {
    return async (dispatch, getState) => {
        let data;
        try {
            data = await client_1.Client4.createAccessControlSyncJob(jobData);
        }
        catch (error) {
            (0, helpers_1.forceLogoutIfNecessary)(error, dispatch, getState);
            return { error };
        }
        return { data };
    };
}
function updateAccessControlPoliciesActive(states, teamId) {
    return (0, helpers_1.bindClientFunc)({
        clientFunc: client_1.Client4.updateAccessControlPoliciesActive,
        params: [states, teamId],
    });
}
